lib/db/structs.go

```diff
@@ -196,98 +195,297 @@ func (vl VersionList) String() string {
+			if j > 0 {
+				b.WriteString(", ")
+			}
+			copy(id[:], dev)
+			fmt.Fprint(&b, id.Short())
+		}
+		fmt.Fprint(&b, "}}")
 	}
 	b.WriteString("}")
 	return b.String()
 }
 
 // update brings the VersionList up to date with file. It returns the updated
-// VersionList, a potentially removed old FileVersion and its index, as well as
-// the index where the new FileVersion was inserted.
-func (vl VersionList) update(folder, device []byte, file protocol.FileInfo, t readOnlyTransaction) (_ VersionList, removedFV FileVersion, removedAt int, insertedAt int, err error) {
-	vl, removedFV, removedAt = vl.pop(device)
-
-	nv := FileVersion{
-		Device:  device,
-		Version: file.Version,
-		Invalid: file.IsInvalid(),
-		Deleted: file.IsDeleted(),
+// VersionList, a device that has the global/newest version, a device that previously
+// had the global/newest version, a boolean indicating if the global version has
+// changed and if any error occurred (only possible in db interaction).
+func (vl VersionList) update(folder, device []byte, file protocol.FileIntf, t readOnlyTransaction) (VersionList, FileVersion, FileVersion, FileVersion, bool, bool, bool, error) {
+	if len(vl.RawVersions) == 0 {
+		nv := newFileVersion(device, file.FileVersion(), file.IsInvalid(), file.IsDeleted())
+		vl.RawVersions = append(vl.RawVersions, nv)
+		return vl, nv, FileVersion{}, FileVersion{}, false, false, true, nil
 	}
+
+	// Get the current global (before updating)
+	oldFV, haveOldGlobal := vl.GetGlobal()
+
+	// Remove ourselves first
+	vl, removedFV, haveRemoved, _, err := vl.pop(folder, device, []byte(file.FileName()), t)
+
+	// Find position and insert the file
+	var added bool
 	i := 0
-	if nv.Invalid {
-		i = sort.Search(len(vl.Versions), func(j int) bool {
-			return vl.Versions[j].Invalid
-		})
-	}
-	for ; i < len(vl.Versions); i++ {
-		switch vl.Versions[i].Version.Compare(file.Version) {
-		case protocol.Equal:
-			fallthrough
-
-		case protocol.Lesser:
-			// The version at this point in the list is equal to or lesser
-			// ("older") than us. We insert ourselves in front of it.
-			vl = vl.insertAt(i, nv)
-			return vl, removedFV, removedAt, i, nil
-
-		case protocol.ConcurrentLesser, protocol.ConcurrentGreater:
-			// The version at this point is in conflict with us. We must pull
-			// the actual file metadata to determine who wins. If we win, we
-			// insert ourselves in front of the loser here. (The "Lesser" and
-			// "Greater" in the condition above is just based on the device
-			// IDs in the version vector, which is not the only thing we use
-			// to determine the winner.)
-			//
-			// A surprise missing file entry here is counted as a win for us.
-			if of, ok, err := t.getFile(folder, vl.Versions[i].Device, []byte(file.Name)); err != nil {
-				return vl, removedFV, removedAt, i, err
-			} else if !ok || file.WinsConflict(of) {
-				vl = vl.insertAt(i, nv)
-				return vl, removedFV, removedAt, i, nil
-			}
+	for ; i < len(vl.RawVersions); i++ {
```

Comment:
Shouldn't this loop be a `func (vl *VersionList) insert(...)`?

Reply:
Done.